[![Project Status: WIP â€“ Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)

# Metagenomics

This hands-on, online course teaches data analysis for metagenomics projects. It is aimed at those with little or no experience of using high performance computing (HPC) for data analysis. In the course we will cover:
- navigating file directories and using the command line
- logging into a remote cloud instance
- using common commands and running analysis programs in the command line
- what is metagenomics?
- following a metagenomics analysis workflow including:
  - performing quality control on reads
  - assembly of reads into a metagenome
  - improving your assembly with polishing
  - binning into species/metagenome-assembled genomes (MAGs)
  - taxonomic assignment and functional annotation using your binned reads

The course is taught as a mixture of live coding, online lectures, self-study and drop-in sessions.

## Prerequisites
This course assumes no prior experience with the tools covered in the workshop but learners are expected to have some familiarity with biological concepts, including the concept of genomes and microbiomes. Participants should bring their own laptops and plan to participate actively.

To get started, follow the directions in the Setup tab to get access to the required software and data for this workshop.

## Acknowledgments
The site infrastructure is based on [The Carpentries](https://carpentries.org/). The first two lessons are based on our [Prenomics](https://cloud-span.github.io/prenomics00-intro/) course; the rest are based on our [Metagenomics](https://cloud-span.github.io/metagenomics00-overview/) course.
