---
title: "Polishing"
#subtitle: "[Metagenomics](/index.qmd)"
# title-block-banner: true
---
<!-- ---
layout: lesson
root: .
--- -->
Welcome back to this **Metagenomics with High Performance Computing** course!

In previous lessons we had an introduction to the command line and completed the first part of our worflow: we assessed the quality of long and short reads and used the long reads to generate a draft assembly. 

Long read data are great for tasks like this because they produce a less fragmented assembly and are more likely to span areas with repeats. However, they are also more likely to contain sequencing errors than short read data.

We must therefore use further tools to improve the quality of our draft assembly. We can "polish" our assembly using both long and short read data. After that, we can perform quality control (QC) checks to see what impact the polishing has had.

By the end of this lesson you will be able to:

- explain what polishing is and why it is important
- polish a draft metagenome assembly with long reads using Medaka 
- polish a draft metagenome assembly with short reads using Pilon
- check the quality of your draft assembly using Seqkit and metaQUAST